app/services/concerns/xccdf/rule_group_rules.rb

```diff
@@ -0,0 +1,105 @@
+# frozen_string_literal: true
+
+module Xccdf
+  # Methods related to saving RuleGroupRules
+  module RuleGroupRules
+    extend ActiveSupport::Concern
+
+    included do
+      def save_rule_group_rules
+        @required_rule_group_rules = rule_groups_with_requires +
+                                     rules_with_requires +
+                                     rule_groups_with_conflicts +
+                                     rules_with_conflicts
+        # Need to remove nil values but compact might not be the most effient
+        @required_rule_group_rules = @required_rule_group_rules.compact
+
+        ::RuleGroupRule.transaction do
+          ::RuleGroupRule.import!(
+            @required_rule_group_rules.select(&:new_record?), ignore: true
+          )
+
+        end
+      end
+
+      private
+
+      def rule_groups_with_requires
+        @rule_groups_with_requires ||= @op_rule_groups.flat_map do |op_rule_group|
+          rule_group_id = rule_group_for(ref_id: op_rule_group.id)&.id
+          op_rule_group.requires.map do |required_ref_id|
+            required_rule_id = rule_for(ref_id: required_ref_id)&.id
+            required_rule_group_id = rule_group_for(ref_id: required_ref_id)&.id
+            if required_rule_id
+              ::RuleGroupRule.find_or_initialize_by(
+                rule_group_id: rule_group_id, required_rule_id: required_rule_id
+              )
+            elsif required_rule_group_id
+              ::RuleGroupRule.find_or_initialize_by(
+                rule_group_id: rule_group_id, required_rule_group_id: required_rule_id
+              )
+            end
+          end
+        end
+      end
+
+      def rules_with_requires
+        @rules_with_requires ||= @op_rules.flat_map do |op_rule|
+          rule_id = rule_for(ref_id: op_rule.id)&.id
+          op_rule.requires.map do |required_ref_id|
+            required_rule_id = rule_for(ref_id: required_ref_id)&.id
+            required_rule_group_id = rule_group_for(ref_id: required_ref_id)&.id
+            if required_rule_id
+              ::RuleGroupRule.find_or_initialize_by(
+                rule_id: rule_id, required_rule_id: required_rule_id
```

Comment:
Where is this `find_or_initialize_by` implemented? I am not able to find it :thinking: also it has a weird method signature, you call it with various arguments across this file. 

This might be the right moment to introduce STI and implement two inherited classes, one for conflicting rule group link and the other for a conflicting rule link in order to handle polymorphism. This is just an idea and it might introduce more problems than it solves though.

Please try to implement the `find_or_initialize_by` method for this class and see how it would work first.

Reply:
I was going off of how it was being used in ProfileRules [here](https://github.com/RedHatInsights/compliance-backend/blob/c4be7550460534166d2074201bcc20330e1c4700/app/services/concerns/xccdf/profile_rules.rb#L28). It looks like from the [find_or_initialize_by documentation](https://apidock.com/rails/v4.0.2/ActiveRecord/Relation/find_or_initialize_by) that it takes a list of attributes of the class